src/evaluator/primitives.ts

```diff
@@ -108,12 +109,13 @@ export const primitive_funcs: Record<string, (...args: EvalValue[]) => EvalResul
     const [arg] = args;
     return ok(sboolean(is_list(arg)));
   },
-  'data?': (...args) => {
+
+  'function?': (...args) => {
     if (args.length !== 1) {
       return err();
     }
     const [arg] = args;
-    return ok(sboolean(is_boxed(arg)));
+    return ok(sboolean(is_boxed(arg) && instanceOfEvalData(arg.val)));
```

Comment:
Requesting for comments? @thomastanck

Reply:
is_boxed(arg) is good enough. refer to the type of EvalValue:

export type EvalValue = SListStruct<EvalData>;

this means that boxed types are EvalData.

You can verify this yourself by asking your ide what the type of arg.val is after is_boxed returns true. Mine says:

![image](https://user-images.githubusercontent.com/829070/108606135-4cf05c80-73f3-11eb-9ef1-34a4549a2bd8.png)

which confirms that it's either a Closure or a Primitive, precisely the type of EvalData:

export type EvalData = Closure | Primitive;

Reply:
Hmm, I was writing this with in mind that we might have other boxed types that are not necessarily functions (which I am guessing that value of `arg.val` could be not `EvalData` anymore?).

So in such cases, we don't want the other boxed type to be part of `function?`, right?

Reply:
no, arg.val will always be EvalData, it just won't be a closure or primitive anymore. furthermore, we have control over what goes in the type EvalData.

in practice what this means is we'll want to make sure that the variant key (continues to) exists on every subtype of EvalData, so that we can use it to discriminate between the different types of EvalData

![image](https://user-images.githubusercontent.com/829070/108607242-2d106700-73fa-11eb-99c1-7be1751bb591.png)

you should note we already have the ability to discriminate between closures and primitives if you check the type

Reply:
ah and, we already make use of it, check the apply function

![image](https://user-images.githubusercontent.com/829070/108607278-55986100-73fa-11eb-9796-878997dbb3a8.png)

Reply:
> no, arg.val will always be EvalData, it just won't be a closure or primitive anymore. furthermore, we have control over what goes in the type EvalData.
>
> in practice what this means is we'll want to make sure that the variant key (continues to) exists on every subtype of EvalData, so that we can use it to discriminate between the different types of EvalData

I see. If that is the case, then would it be better if in the `function?` primitive I check that the variant of value is of type `Closure` or `Primitive`?

In the flavor of:
```
is_boxed(arg) && (arg.val.variant === EvalDataType.Closure || arg.val.variant === EvalDataType.Primitive)
```

Reply:
yeah that'd work

Reply:
Ok. I made the change in 152d096